components/bitcoind/src/utils/redis_notifier.rs

```diff
@@ -0,0 +1,362 @@
+    pub fn new(redis: &RedisConfig) -> Result<Self, String> {
+        let client = Client::open(redis.url.as_str())
+            .map_err(|e| format!("unable to connect to redis: {e}"))?;
+        Ok(Self {
+            queue: redis.queue.clone(),
+            client,
+        })
+    }
+
+    fn network_str(bitcoind: &BitcoindConfig) -> &'static str {
+        match bitcoind.network {
+            bitcoin::Network::Bitcoin => "mainnet",
+            bitcoin::Network::Testnet | bitcoin::Network::Testnet4 => "testnet",
+            bitcoin::Network::Signet => "signet",
+            bitcoin::Network::Regtest => "regtest",
+            _ => "unknown",
+        }
+    }
+
+    pub async fn notify(
+        &self,
+        indexer: &str,
+        bitcoind: &BitcoindConfig,
+        apply_blocks: &[BlockIdentifier],
+        rollback_blocks: &[BlockIdentifier],
+    ) -> Result<(), String> {
+        let payload = build_message_json(indexer, bitcoind, apply_blocks, rollback_blocks)?;
+        // In-test capture path (no real Redis required)
+        if std::env::var("REDIS_TEST_CAPTURE").ok().as_deref() == Some("1") {
+            capture_messages().lock().unwrap().push(payload);
+            return Ok(());
+        }
+
+        let mut conn = self
+            .client
+            .get_multiplexed_tokio_connection()
+            .await
+            .map_err(|e| format!("unable to get redis connection: {e}"))?;
+        conn.rpush::<_, _, ()>(&self.queue, payload)
+            .await
+            .map_err(|e| format!("unable to rpush redis message: {e}"))?;
+        Ok(())
+    }
+}
+
+static CAPTURED_MESSAGES: OnceLock<Mutex<Vec<String>>> = OnceLock::new();
+
+fn capture_messages() -> &'static Mutex<Vec<String>> {
+    CAPTURED_MESSAGES.get_or_init(|| Mutex::new(Vec::new()))
+}
+
+#[cfg(test)]
+fn take_captured_messages() -> Vec<String> {
+    let mut guard = capture_messages().lock().unwrap();
+    let messages = guard.clone();
+    guard.clear();
+    messages
+}
+
+fn js_sys_time_ms() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .map(|d| d.as_millis())
+        .unwrap_or(0)
+}
+
+pub fn build_message_json(
+    indexer: &str,
+    bitcoind: &BitcoindConfig,
+    apply_blocks: &[BlockIdentifier],
+    rollback_blocks: &[BlockIdentifier],
+) -> Result<String, String> {
+    let network = RedisNotifier::network_str(bitcoind);
+    let id = format!(
+        "bitcoin-{}-{}-{}",
+        indexer,
+        apply_blocks
+            .last()
+            .map(|b| b.index.to_string())
+            .unwrap_or_else(|| "na".to_string()),
+        js_sys_time_ms()
+    );
+    let message = Message {
+        id,
+        payload: IndexProgressPayload {
+            chain: "bitcoin",
+            network,
+            indexer,
+            apply_blocks: apply_blocks
+                .iter()
+                .map(|b| BlockIndexRef {
+                    hash: b.hash.clone(),
```

Comment:
we need to make sure these block hashes start with `0x`, can you double check if this is the case?